wi-fi: support STA mode

julianuziemblo opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR adds STA mode to Wi-Fi manager. For easier Wi-Fi management, the `/dev/wifi` device had to be split into three devices:
- `/dev/wifi/ctrl`: Wi-Fi state management: on/off. Putting Wi-Fi "on" initializes the WHD interfaces: STA as primary and AP as secondary. It also creates the latter 2 devices for managing STA and AP modes.
- `/dev/wifi/sta`: STA mode management: set idle timeout, set SSID and key of the network you want to connect to, connect/disconnect.
- `/dev/wifi/ap`: AP mode management: set idle timeout, set SSID and key of the access point, start/stop (API unchanged)

Idle timeout should always be set `!= 0` in both modes as it is used to periodically check if the bus is still operational (the bus can fail if e.g. device gets disconnected).

This PR also adds by default, concurrent mode (AP+STA together at the same time) is enabled. To disable it, define WIFI_CONCURRENT_MODE_ENABLED to 0 in `lwipopts.h`.

This PR also allows setting AP static address by writing to `/dev/wifi/ap`:
- "ip_addr a.b.c.d" for IP address
- "netmask a.b.c.d" for netmask
- "gateway a.b.c.d" for gateway address
Written address is only validated if it itself is valid. It is
not checked for being a valid interface/netmask/gateway address.
This check is only done on AP start.

This PR also also adds IP reporting on read:
- AP reports its static IP, netmask, gateway
- STA reports its dynamic IP, netmask, gateway

## Motivation and Context
STA mode (client mode) needed on some platforms (especially imxrt).

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change):
    - `/dev/wifi` device split into three devices (described in "Description" section above)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `armv7m7-imxrt117x-evk`,  `armv7a7-imx6ull-dataro`, NILEE.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work:
    - https://github.com/phoenix-rtos/phoenix-rtos-lwip/pull/118 - *maybe not necessary, but needed to run on imrt117x with Sterling LWB5+ USB dongle*
    - https://github.com/phoenix-rtos/libphoenix/pull/466
- [ ] I will merge this PR by myself when appropriate.

julianuziemblo committed:
wi-fi/main: MISRA fixes

JIRA: RTOS-1166

julianuziemblo committed:
wi-fi/main: init wifi interface in separate function

JIRA: RTOS-1166

julianuziemblo committed:
wi-fi/main: trim input before parsing

cmd-line utilities may pass extra space or newline as opposed to NULL
byte at the end of the string, making them hold wrong values.

JIRA: RTOS-1166

julianuziemblo committed:
!wi-fi/main: add support for STA mode

This change required some reorganising of the code.
3 new devices were added in /dev/wifi/:
- /dev/wifi/ctrl: control (Wi-Fi on/off)
- /dev/wifi/ap:   AP mode ctrl (set credentials, start/stop)
- /dev/wifi/sta:  STA mode ctrl (set credsentials, connect/disconnect)

Migration for AP:
- turn on Wi-Fi by writing "on" to /dev/wifi/ctrl
- set credentials, timeout and start/stop like before, writing
  to /dev/wifi/ap instead of /dev/wifi

Using STA:
- turn on Wi-Fi by writing "on" to /dev/wifi/ctrl
- set credentials the same as in AP mode, writing them to /dev/wifi/sta
- write "connect"/"disconnect" to /dev/wifi/sta to manage connection
  to an access point

Idle timeout should always be set `!= 0` in both modes as it is used
to periodically check if the bus is still operational (the bus can fail
if e.g. USB device gets disconnected).

JIRA: RTOS-1166

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +316,576 @@ static void wifi_ap_thread(void *arg)
+	if (result_ptr != NULL) {
+		*result_ptr = NULL;
+	}
+	mutexUnlock(wifi_common.dev[STA_DEV_ID].lock);
```
WHD internally passes &scan_result as pointer, so it IS passed as `&(&scan_result)` anyway. Making it NULL just signals to WHD to stop scanning as we've encountered the appropriate result.

julianuziemblo commented:
/gemini review

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -94,15 +95,23 @@ static int wifi_ap_set_idle_timeout(const char *data, size_t len)
-	if (timeout < 0) {
+	errno = 0;
+	timeout = strtol(buf, &endp, 0);
+	if (errno != 0 || endp == buf || timeout == LONG_MAX) {
```
`timeout == LONG_MAX` should be already handled by `errno != 0` (ERANGE).

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -121,13 +121,33 @@ static int wifi_ap_set_idle_timeout(const char *data, size_t len)
 		return -1;
 	}
 
+	ssid = trim(ssid, &len);
```
Move it before SSID size check.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -149,6 +169,8 @@ static int wifi_ap_set_key(const char *key, size_t len)
 		return -1;
 	}
 
+	key = trim(key, &len);
```
Move it before PSK size checks.

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -446,6 +468,7 @@ static int wifi_dev_read(char *data, size_t size, off_t offset)
 
 static int wifi_dev_write(const char *data, size_t size)
 {
+	data = trim(data, &size);
```
I don't think it will as before `strncmp` there is `size >= 8` check which will not be passed, hence not calling `wifi_ap_set_idle_timeout`

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -53,25 +65,42 @@
 	struct {
 		bool busy;
-		char buf[128];
+		char buf[16];
```
You can use move it to stack since it's very small.

julianuziemblo replied:
oh yeah, I forgot it matters here, my bad.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -53,25 +65,41 @@
+			struct wifi_device sta;
+			struct wifi_device ap;
+		};
+		struct wifi_device dev[2];
```
Assuming that `CY_LWIP_AP_NW_INTERFACE` is 0 and `CY_LWIP_STA_NW_INTERFACE` is 1 is dangerous.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -158,29 +219,31 @@ static int wifi_ap_is_idle(void)
 
 
-static void wifi_ap_main_loop(void)
+static void wifi_ap_main_loop(void) 
```
Remove extra space at the end of line.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -53,25 +65,41 @@
+	atomic_bool initialized;
+	handle_t tid;
+
+	volatile uint8_t flags;
```
`flags` is used by AP only and protected by `ap.lock` so you should move it to `wifi_device` struct.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +315,583 @@ static void wifi_ap_thread(void *arg)
-			/* FIXME: temporary workaround - find out why AP doesn't start */
-			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "retrying to start Wi-Fi AP\n");
+	else if (id == STA_DEV_ID) {
+		if (size == (sizeof("connect") - 1) && strncmp("connect", data, size) == 0) {
```
Consider moving `id == STA_DEV_ID` and `id == AP_DEV_ID` checks to command `if` condition to avoid repeating error message.

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +315,583 @@ static void wifi_ap_thread(void *arg)
+	time_t when, now, timeout_remaining;
+	uint32_t scan_timeout;
+
+	mutexLock(wifi_common.sta.lock);
```
I now refactored the STA to work in similar fashion to AP: it spawns its own thread and periodically calls "wifi_is_idle" (`wifi_sta_is_connected` in this case) to check if the connection is still present. All checks are now performed the same for both wifi devices in `wifi_dev_start()` and `wifi_dev_stop()`.

julianuziemblo replied:
I also added `!wifi_common.initialized` check to the msgthread

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +357,590 @@ static void wifi_ap_thread(void *arg)
 	}
-	else if (strncmp("start", data, size) == 0) {
-		unsigned int retries = WIFI_START_RETRIES;
+	else if (id == STA_DEV_ID && size == (sizeof("connect") - 1) && strncmp("connect", data, size) == 0) {
```
Consider adding `STRLEN` macro `(sizeof(s) - 1)`, since this pattern is repeated several times.

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +357,590 @@ static void wifi_ap_thread(void *arg)
+			(void)whd_wifi_stop_scan(wifi_common.sta.iface.whd_iface);
+
+			if (scan_timeout != 0) {
+				now = time(NULL);
```
okay, good catch then - did not think about it

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -221,234 +357,590 @@ static void wifi_ap_thread(void *arg)
+
+			if (size >= (sizeof("on") - 1) && strncmp("on", data, size) == 0) {
+				oid_t oid __attribute__((unused));
+				if (wifi_common.initialized || lookup(AP_DEV_NAME, NULL, &oid) >= 0 || lookup(STA_DEV_NAME, NULL, &oid) >= 0) {
```
there's no way to wait at this point - maybe failing here is a good thing after all?

`sys_thread_join` does not work - as I said, we cannot wait due to sequential handling of messages. The best we can do is fail with a "too soon" message, as is done for now.

julianuziemblo committed:
wi-fi/main: allow setting AP static IP + report IP address on device read

Allow setting AP static address by writing to /dev/wifi/ap:
- "ipaddr a.b.c.d" for IP address
- "netmask a.b.c.d" for netmask
- "gateway a.b.c.d" for gateway address
Written address is only validated if it itself is valid. It is
not checked for being a valid interface/netmask/gateway address.
This check is only done on AP start.

IP reporting:
- AP reports its static IP, netmask, gateway
- STA reports its dynamic IP, netmask, gateway

JIRA: RTOS-1166

Darchiv commented on wi-fi/lwip/main.c:
```diff
@@ -121,8 +121,28 @@ static int wifi_ap_set_idle_timeout(const char *data, size_t len)
+static const char *trim(const char *str, size_t *len)
+{
+	size_t sz = *len;
+	while (sz > 0 && !isgraph(*str)) {
```
Unfortunately, `isgraph()` takes only ASCII printable characters into account and ignores locale. All Unicode characters (UTF-8 encoded as >=0x80) are trimmed. For future reference.

Is WHD/chip even capable of setting non-ASCII SSIDs?

julianuziemblo replied:
It's technically possible as `whd_ssid_t` is `length` and `value`, but I don't know if we wanna support it out-of-the-box. I'm also not sure if the chips support it. What do you think: should anything be done here?

ziemleszcz commented:
/gemini review

julianuziemblo commented on wi-fi/lwip/main.c:
```diff
@@ -457,9 +1028,9 @@ static void wifi_msg_thread(void *arg)
 	unsigned int port;
 
-	result = cy_log_init(CY_LOG_INFO);
+	result = cy_log_init(CY_LOG_DEBUG);
```
uh, missed this one

Darchiv commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,806 @@ static int wifi_ap_set_key(const char *key, size_t len)
+}
+
+
+static bool wifi_sta_state(whd_ssid_t *ssid)
```
Wrong return value. Should be `enum wifi_state`

Darchiv commented on wi-fi/lwip/main.c:
```diff
@@ -80,56 +127,89 @@ static ip_static_addr_t ap_addr = {
 	.gateway = IPADDR4_INIT_BYTES(192, 168, 2, 1)
 };
 
+static whd_scan_result_t scanResult;
```
Why store it outside of `wifi_common`?

julianuziemblo replied:
mimicked `ap_addr`. I'll transfer it to common with `ap_addr` as we're making static address change possible.

Darchiv commented on wi-fi/lwip/main.c:
```diff
@@ -80,56 +127,89 @@ static ip_static_addr_t ap_addr = {
 	.gateway = IPADDR4_INIT_BYTES(192, 168, 2, 1)
 };
```
So AP network IPs are hardcoded? Optimally, these should be settable via `/dev/wifi/ap`. If out of scope, then please mark these with a TODO and `const` (if possible)

julianuziemblo replied:
how about now?

Darchiv reviewed (commented):
Left some remarks.

Is /dev/wifi/* interface documented somewhere? I mean, you have written what has changed in this PR, but is there a single point of reference for these devices?

Darchiv commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,806 @@ static int wifi_ap_set_key(const char *key, size_t len)
 
 
-static int wifi_ap_is_idle(void)
+static bool wifi_ap_state(void)
```
Wrong return value. Should be `enum wifi_state`

julianuziemblo commented:
> Is /dev/wifi/* interface documented somewhere? I mean, you have written what has changed in this PR, but is there a single point of reference for these devices?

AFAIK: no, there isn't. Maybe I could create document on YT and then we could consider adding it to our documentation?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +222,827 @@ static int wifi_ap_set_key(const char *key, size_t len)
-			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
-			cybsp_free();
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't get network's security\n");
 			break;
```
Unlocked `wifi_common.sta.lock`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +222,827 @@ static int wifi_ap_set_key(const char *key, size_t len)
-			cybsp_wifi_deinit(wifi_common.iface.whd_iface);
-			cybsp_free();
+			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't join requested network\n");
 			break;
```
Unlocked `wifi_common.sta.lock`.

julianuziemblo replied:
good eye, thx

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +222,827 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		cybsp_wifi_deinit(wifi_common.sta.iface.whd_iface);
+		cybsp_free();
+	}
+	else if (wifi_common.ap.iface.whd_iface != NULL) {
```
Why `else if` instead of `if`?

julianuziemblo replied:
cybsp_wifi_deinit() deinitializes both interfaces at once; the iface pointer is just to retrieve the driver. I can add comment here, it's commented like that in cybsp_wifi_deinit definition. What do you think?

ziemleszcz replied:
Please add some comment.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +222,827 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		wifi_dev_stop(&wifi_common.ap);
+	}
+	else if (id == AP_DEV_ID && size >= CONST_STRLEN("ip_addr ") && strncmp("ip_addr ", data, 8) == 0) {
+		if (ip4addr_aton(data + 8, &wifi_common.apAddr.addr) != 0) {
```
Note that `data + 8` may be not NULL terminated string.

julianuziemblo replied:
Do you think we should terminate it here somehow, or just let `ip4addr_aton` do its thing?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +222,827 @@ static int wifi_ap_set_key(const char *key, size_t len)
+			off_t offset = msg->i.io.offs;
+			int cnt = min(msg->o.size, wifi_common.ctrl.len - offset);
+
+			if (offset > wifi_common.ctrl.len) {
```
Move it before `cnt` calculation.

julianuziemblo commented:
/gemini review

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +224,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		wifi_dev_stop(&wifi_common.ap);
+	}
+	else if (id == AP_DEV_ID && size >= CONST_STRLEN("ip_addr ") && strncmp("ip_addr ", data, 8) == 0) {
+		return wifi_write_addr(&wifi_common.apAddr.addr, data + 8, size - 8);
```
I would follow `wifi_set_` name convention for these functions.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -53,83 +74,146 @@
-static int wifi_ap_set_ssid(const char *ssid, size_t len)
+static int wifi_set_ssid(id_t id, const char *ssid, size_t len)
 {
+	ssid = trim(ssid, &len);
```
Input is already trimmed.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -53,83 +74,146 @@
-static int wifi_ap_set_key(const char *key, size_t len)
+static int wifi_set_key(id_t id, const char *key, size_t len)
 {
+	key = trim(key, &len);
```
Input is already trimmed.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	}
+
+#if !WIFI_CONCURRENT_MODE_ENABLED
+	if (other->tid != 0) {
```
`tid` check should be done with `dev->lock` locked.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	do {
+		enum wifi_state state = wifi_sta_state(&connectedTo);
+		if (state == wifi_running) {
+			if (memcmp(&connectedTo, &ssid, sizeof(ssid)) == 0) {
```
You should compare `ssid.value` using `ssid.length`. Additionally, verify that the lengths are equal, since one SSID could be a prefix of another.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		case WHD_SCAN_INCOMPLETE: {
+			if ((*result_ptr)->SSID.length != 0) {
+				whd_ssid_t *ssid = user_data;
+				if (memcmp((*result_ptr)->SSID.value, ssid->value, ssid->length) != 0) {
```
Verify that the lengths are equal, since one SSID could be a prefix of another.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+	do {
+		enum wifi_state state = wifi_sta_state(&connectedTo);
+		if (state == wifi_running) {
```
Why would STA be running here?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+			if (memcmp(&connectedTo, &ssid, sizeof(ssid)) == 0) {
+				wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_INFO, "already connected to this network\n");
+				result = WHD_SUCCESS;
+				break;
```
Result is success but ` wifiJoined` and `ifaceAdded` remain false.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,315 +220,863 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	whd_ssid_t ssid, connectedTo;
+	uint8_t key[WSEC_MAX_PSK_LEN];
+	uint8_t keyLen;
+	time_t when, now, left;
```
Declare `when` and `now` only if  `STA_SCAN_TIMEOUT_SECONDS != 0`?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +221,937 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		wifi_main_loop(&wifi_common.ap);
+	}
+
+	whd_wifi_stop_ap(wifi_common.ap.iface.whd_iface);
```
The AP is being stopped unconditionally, even if it may not have been started.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +221,937 @@ static int wifi_ap_set_key(const char *key, size_t len)
+					return size;
+				}
+
+				wifi_deinit_interfaces();
```
`_wifi_deinit()` already calls `wifi_deinit_interfaces()`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +221,937 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+		mutexLock(wifi_common.lock);
+		/* safe because we already set our thread's tid to 0 */
+		_wifi_deinit();
```
Why are you removing AP and STA devices? This is equivalent to running `off` command. Also note that `wifi_common.initialized` is not set to `false`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +221,937 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+		mutexLock(wifi_common.lock);
+		/* safe because we already set our thread's tid to 0 */
+		_wifi_deinit();
```
Why are you removing AP and STA devices? This is equivalent to running `off` command. Also note that `wifi_common.initialized` is not set to `false`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +221,937 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	if (id == STA_DEV_ID) {
+		SNPRINTF_APPEND(overflow, buf, size, "ssid=%.*s\n", wifi_common.sta.ssid.length, wifi_common.sta.ssid.value);
+		whd_ssid_t ssid;
+		enum wifi_state state = wifi_sta_state(&ssid);
```
There is no `(dev->iface.whd_iface == NULL || whd_wifi_get_mac_address(dev->iface.whd_iface, &mac) == WHD_BUS_FAIL)` check as in `wifi_get_state()`.

julianuziemblo committed:
wi-fi: add way for WHD Bus Layer to report failure

When Bus Layer fails, we invoke registered error handlers
and exit from the WHD Thread. Bus Layer should report
WHD_BUS_FAIL on read-write only on unrecoverable fails.

In USB Bus, this is achieved by checking if device is still
connected on read error. If not - we report bus fail.

JIRA: RTOS-1166

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +233,990 @@ static int wifi_ap_set_key(const char *key, size_t len)
+static void _wifi_ctrl_action(enum wifi_action action)
 {
-	oid_t wifi_oid = { port, WIFI_DEV_ID };
+	wifi_common.ctrl.action = action;
```
There is no protection against overwriting an ongoing action. Consider waiting for the previous action to complete before setting a new one.

julianuziemblo replied:
Performing and setting an action is under a mutex, shouldn't that suffice?

ziemleszcz replied:
I'm not sure this is safe. What if two threads attempt to set the action concurrently? The first thread may set it, but could the second thread acquire the mutex before `condWait()` in `wifi_ctrl_thread()`?

julianuziemblo replied:
But the ctrl thread that performs the action is under the ctrl mutex, and only releases it after handling the action - in condWait. I'm not sure how other threads could acquire the mutex. Is it possible for another than ctrl thread to be scheduled when we condSignal?

ziemleszcz replied:
`Is it possible for another than ctrl thread to be scheduled when we condSignal?`: that's exactly what I'm worried about. `condWait` releases and reacquires the mutex. Anyway this probably implementation specific.

julianuziemblo replied:
the only solution I see is making it atomic and waiting until the operation is complete - do you have better ideas or do you think this one will suffice?

ziemleszcz replied:
Crazy idea: convert `wifi_ctrl_thread` into a full message handler (with a separate port for `/dev/wifi/ctrl`) and handle actions synchronously using the message API.
Alternatively, you could wait for the action to complete, similar to `wifi_dev_start()`, using a condition variable.
Handling `start` command synchronously would be preferable from my perspective.

julianuziemblo replied:
hm, sounds reasonable - will give it a try

ziemleszcz commented on wi-fi/whd/bus_protocols/whd_bus_usb_protocol.c:
```diff
@@ -687,7 +694,7 @@ whd_result_t whd_bus_usb_read_frame(whd_driver_t whd_driver, whd_buffer_t* buffe
     }
 
-    return 1;
+    return WHD_BUS_FAIL;
```
Why not `WHD_NO_PACKET_TO_RECEIVE`?

ziemleszcz commented on wi-fi/whd/bus_protocols/whd_bus_usb_protocol.c:
```diff
@@ -603,6 +606,10 @@ static whd_bool_t whd_bus_usb_wake_interrupt_present(whd_driver_t whd_driver)
 static uint32_t whd_bus_usb_packet_available_to_read(whd_driver_t whd_driver)
 {
+    if (whd_bus_usb_common.rx.failed) {
+        return WHD_BUS_FAIL;
```
It seems that this function doesn't return an error code but number of RX packets.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +233,990 @@ static int wifi_ap_set_key(const char *key, size_t len)
+static enum wifi_state wifi_get_state(struct wifi_device *dev)
+{
+	if (dev->iface.whd_iface == NULL) {
+		return wifi_idle;
```
This should be a different state (`wifi_off`?). Originally, there was only `wifi_ap_is_idle()` function, which returned true if no client was connected to the AP and false otherwise. So, `wifi_idle` state should not be equivalent to having no AP at all.
I'm not sure how this relates to STA mode. For AP mode, we shut it down if it remains idle for `timeout` seconds. The idea was that clients can connect to DATARO AP only for a limited time after it starts (security requirement).
Do we need a similar timeout for STA mode? If so, what is its intended purpose?

julianuziemblo replied:
The thread's purpose for STA is to check if we're still connected (e.g. the network seized to exist). You might be right it shouldn't then be able to be configured to have an infinite timeout (0). I'll try to revise it for STA.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +233,990 @@ static int wifi_ap_set_key(const char *key, size_t len)
-		mutexLock(wifi_common.lock);
-		condWait(wifi_common.cond, wifi_common.lock, cond_timeout * 1000000ULL);
+		mutexLock(dev->lock);
+		condWait(dev->cond, dev->lock, condTimeout * 1000000ULL);
```
This appears to break AP `timeout` feature. `condTimeout` is 0 when the loop starts, resulting in an indefinite wait that is only terminated by `stop` command.
As a result, if AP `timeout` is set after the loop has started, it will never take effect.

julianuziemblo replied:
yes - and that's how it was in the original code. The logic was not changed. Would you like it to work the way you describe it?

ziemleszcz replied:
OK, this is fine. It might be worth adding a comment that the timeout must be set before starting the AP.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,391 +233,990 @@ static int wifi_ap_set_key(const char *key, size_t len)
+		unsigned int retries = AP_START_RETRIES;
 
-		while (wifi_ap_start() < 0 && retries-- > 0) {
+		while (wifi_dev_start(&wifi_common.ap) < 0 && retries-- > 0) {
```
Probably no longer needed after last changes by jmaksymowicz.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,930 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	if (wifi_common.msgport != -1) {
+		portDestroy(wifi_common.msgport);
+		/* only wait for thread if we can signalize for it to close with portDestroy */
+		if (wifi_common.msgtid != 0) {
```
`msgtid` is not set to the actual `TID`.

ziemleszcz replied:
`msgport` is not set to 0 if `wifi_common.msgtid == 0` (created msg thread failed).

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,930 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	} while (0);
+
+	/* set this regardless as it will be set to false in _wifi_deinit if initialization failed */
+	wifi_common.initialized = true;
```
Unsafe, `wifi_msg_thread` could try to handle some message when initialization failed.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,930 @@ static int wifi_ap_set_key(const char *key, size_t len)
 
-	if (portCreate(&port) < 0) {
-		wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't create Wi-Fi port\n");
+	if (wifi_common.msgport != -1) {
```
`msgport` is unsigned.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,930 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+	int err;
+
+	wifi_common.msgport = -1;
```
`msgport` is unsigned.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,930 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	}
+
+	wifi_deinit_interfaces();
+	if (wifi_common.sta.lock != -1) {
```
What if somebody reads `/dev/wifi/ap` or `/dev/wifi/sta` at this point? These devices are destroyed at the end of this function.

julianuziemblo replied:
We've already joined the msgthread by then, so that's not possible

julianuziemblo replied:
Should we first delete the devices and then join the msgthread?

ziemleszcz replied:
OK, you are right.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +229,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	else if (size >= CONST_STRLEN("key ") && strncmp("key ", data, 4) == 0) {
+		wifi_set_key(id, data + 4, size - 4);
+	}
+	else if (id == STA_DEV_ID && size == CONST_STRLEN("connect") && strncmp("connect", data, size) == 0) {
```
I'm not sure about a synchronous STA connection. How long could it block if no network is available? This operation might be better handled asynchronously.

julianuziemblo replied:
It will block no more than STA_TIMEOUT seconds defined at the top of the file

julianuziemblo replied:
... But I guess I made it configurable, so maybe it should be async

ziemleszcz replied:
OK, I'll leave the decision to you. For 10 seconds, it could remain synchronous. My main concern was preserving synchronous AP operations.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+	if (err < 0) {
+		_wifi_deinit();
+	}
```
`else wifi_common.initialized = true`?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	destroy_dev(AP_DEV_NAME);
+	destroy_dev(STA_DEV_NAME);
+
+	if (wifi_common.msgport != (unsigned int)-1) {
```
Nitpick: `(uint32_t)-1`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+	if (wifi_common.msgport != (unsigned int)-1) {
+		portDestroy(wifi_common.msgport);
+		wifi_common.msgport = (unsigned int)-1;
```
Nitpick: `(uint32_t)-1`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+{
+	int err;
+
+	wifi_common.msgport = (unsigned int)-1;
```
Nitpick: `(uint32_t)-1`.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+					msg.o.err = -EINVAL;
+					break;
+				}
+				msg.o.err = size;
```
` = msg.i.size`?

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+				if (size == CONST_STRLEN("on") && strncmp("on", data, size) == 0) {
+					if (wifi_init() < 0) {
+						wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "failed to initialize Wi-Fi\n");
```
Not true if Wi-Fi is already intialized. Maybe move this log to `wifi_init()` or return 0 in `wifi_init()` for this case.

julianuziemblo replied:
technically it is true then :smile: but I see what you mean, will move it

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,944 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	else if (id == AP_DEV_ID && size == CONST_STRLEN("stop") && strncmp("stop", data, size) == 0) {
 		wifi_ap_stop();
 	}
+	else if (id == AP_DEV_ID && size >= CONST_STRLEN("ipaddr ") && strncmp("ipaddr ", data, 8) == 0) {
```
8 -> 7

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,949 @@ static int wifi_ap_set_key(const char *key, size_t len)
-		result = cybsp_init();
-		if (result != CY_RSLT_SUCCESS) {
-			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't init Wi-Fi HW\n");
+		if (wifi_add_interface(&wifi_common.ap.iface, &wifi_common.ap.apPriv.addr) < 0) {
```
Missing `whd_wifi_stop_ap()` call on failure.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,949 @@ static int wifi_ap_set_key(const char *key, size_t len)
+			whd_ssid_t ssid;
+			(void)wifi_sta_get_connected(&ssid);
+			struct netif *netif = cy_lwip_get_interface(wifi_common.dev[id].iface.role);
+			SNPRINTF_APPEND(overflow, buf, size, "current=%.*s\n", ssid.length, ssid.value);
```
`ssid` is uninitialized on `wifi_sta_get_connected()` failure.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,949 @@ static int wifi_ap_set_key(const char *key, size_t len)
+
+			if (clock_gettime(CLOCK_MONOTONIC, &tp) < 0) {
+				wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "couldn't get current time: %s (%d)\n", strerror(errno), errno);
+				break;
```
result = WHD_SCAN_ABORTED

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,958 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	} while (0);
+
+	mutexLock(wifi_common.ap.lock);
+	wifi_common.ap.apPriv.flags |= started ? WIFI_FLAG_STARTED : WIFI_FLAG_FAILED;
```
started -> result == WHD_SUCCESS

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,958 @@ static int wifi_ap_set_key(const char *key, size_t len)
-		result = cybsp_init();
-		if (result != CY_RSLT_SUCCESS) {
-			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "can't init Wi-Fi HW\n");
+		if (wifi_add_interface(&wifi_common.ap.iface, &wifi_common.ap.apPriv.addr) < 0) {
```
Suggestion: `wifi_add_interface()` could return `whd_result_t` directly.

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,957 @@ static int wifi_ap_set_key(const char *key, size_t len)
-			wm_cy_log_msg(CYLF_MIDDLEWARE, CY_LOG_ERR, "retrying to start Wi-Fi AP\n");
-		}
+
+static int wifi_dev_write(id_t id, const char *data, size_t size)
```
This function should return back exactly `size` on success (not trimmed size).

ziemleszcz commented on wi-fi/lwip/main.c:
```diff
@@ -140,369 +238,957 @@ static int wifi_ap_set_key(const char *key, size_t len)
+	else if (id == AP_DEV_ID) {
+		SNPRINTF_APPEND(overflow, buf, size, "timeout=%u\n", wifi_common.ap.apPriv.timeout);
+		SNPRINTF_APPEND(overflow, buf, size, "ssid=%.*s\n", wifi_common.ap.ssid.length, wifi_common.ap.ssid.value);
+		SNPRINTF_APPEND(overflow, buf, size, "ip_addr=%s\n", ipaddr_ntoa(&wifi_common.ap.apPriv.addr.addr));
```
`ipaddr` is used for STA.
